### ISSUE TITLE
Sik-Ho Tang | Review: Stochastic Depth (Image Classification).

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#104]
Sik-Ho Tang. [Review: Stochastic Depth (Image Classification)](https://towardsdatascience.com/review-stochastic-depth-image-classification-a4e225807f4a).

**Comment by NorbertZheng** [comment]
## Overview
In this story, Stochastic Depth is shortly reviewed. Stochastic Depth, **a training procedure to train short networks and use deep networks at test time**. This is motivated by the dilemma:
 - Using Deep Model: We can get better prediction accuracy but it is difficult to train due to gradient vanishing problem.
 - Using shallower Model: We can train the network more easily but the prediction accuracy sometimes is not good enough.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/d544476f-828c-49f0-90a6-031d8ab25e15)
Can we?

By using Stochastic Depth, the network is shorten during training, i.e.
 - **a subset of layers is randomly dropped and bypass them with the identity function**.

And a full network is used during testing/inference. By this mean:
 - Training time is reduced substantially.
 - Test error is improved significantly as well.

This is a paper in 2016 ECCV with about 400 citations while I was writing story.

**Comment by NorbertZheng** [comment]
## Brief Revision of Original [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8)
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/475dc952-f0bb-49ac-98d9-01951328cf92)
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/c718acef-4d05-453a-911a-9f8da704f36b)
One $l$-th Residual Block (ResBlock) in Original ResNet.

In Original [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8), suppose $H_{l-1}$ is the input of the above ResBlock, $H_{l-1}$ will go through two paths.
 - Upper Path $f_{l}(H_{l-1})$: Conv > BN > ReLU > Conv > BN
 - Lower Path $id(H_{l-1})$: Identity path without modification of the input

Then the outputs of these two paths are added together, then ReLU, and become $H_{l}$.

By using the identity path, i.e. the skip connection or shortcut connection, we can keep the input signal, and attempt to avoid the gradient vanishing problem. Finally, we can obtain a very deep model.

However, training time of such kind of deep models is long.

Also, there might be overfitting problem.

**Comment by NorbertZheng** [comment]
## [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8) with Stochastic Depth
### Training
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/0f9325c9-c926-4af7-9998-c065cbe6e7f1)
Some ResBlocks are Dropped Randomly Based on Bernoulli Random Variable.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/ebd3845e-d065-4470-bf2a-4f1c1ed133e5)
Linear Decay Rule.

By using Stochastic Depth, during training, for each mini-batch, each ResBlock would have a “survival” probability $p_{l}$.

It is kept if survived. Otherwise, it is skipped as shown above.

 - **Networks trained with Stochastic Depth can be interpreted as an implicit ensemble of networks of different depths.**

To decide $p_{l}$,
 - One is to have constant value of $p_{l}$ along the whole model.
 - One is to have linear decay rule for $p_{l}$ along the whole model.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/6971becb-b134-44cb-b68c-948ce3b62bb8)
If linear decay rule is used, earlier (deeper) layers have larger chance to be survived (skipped) as exampled above.

During training, the expected network depth is shorter than the whole network depth. If the whole network depth $L=110$, $p_{L}=0.5$, the expected network depth $E(L’)=40$. Therefore, training time is much shorter.

**Comment by NorbertZheng** [comment]
### Testing
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/37ab828e-67d6-4286-b712-0f4f949578b4)
Weighted by $p_{l}$ during testing.

Stochastic depth during testing requires small modifications to the network. Since, during training, functions $f_{l}$ are only active for a fraction $p_{l}$ of all updates, and the corresponding weights of the next layer are calibrated for this survival probability. We therefore need to re-calibrate the outputs of any given function $f_{l}$ by the expected number of times it participates in training, $p_{l}$.

**Comment by NorbertZheng** [comment]
Note: There is down-weighting!!!

**Comment by NorbertZheng** [comment]
## Some Results and Analysis
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/944643a2-783d-4bc8-87be-c0a95529cc5f)
CIFAR10, CIFAR100, ImageNet Results.

### CIFAR-10, CIFAR-100, SHVN
For example at top left of the figure:
 - Training loss of constant depth < Training loss of Stochastic Depth.
 - Test loss of constant depth (6.41%) > Test loss of Stochastic Depth (5.25%).
 - **That means Stochastic Depth has reduced overfitting.**

Similar trend for CIFAR-100 and SHVN using either 110 or 1202-Layer model.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/feed86b2-7ca4-4fe3-9a95-bd861acfff4d)
Training Time.

 - Training time is much shorter as well.

**Comment by NorbertZheng** [comment]
### ImageNet
For the ImageNet at the bottom right of the figure:
 - Though the validation error is very close (23.06% for constant depth and 23.38% for Stochastic Depth), training time is 25% shorter.
 - If the training time is equal, this reaches a final error of 21.98%.

Below is the detailed results for each dataset:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/dc34f0d1-9b95-42f1-a6e2-6b4d68f93201)
Detailed Results for Each Dataset.

**Comment by NorbertZheng** [comment]
### Analyses
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/8bea306d-7be1-477c-872e-8942c08fc7b9)
Mean Gradient Magnitude.

By looking at Mean Gradient Magnitude for each epoch, Stochastic Depth has consistently larger weights than constant depth. That means, gradient vanishing problem is less severe in Stochastic Depth.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/8deedd1e-b68a-49bc-9ef8-326bef0cbf7b)
Test Error vs Survival Probability (Left), and Test Error Heatmap with P_L vs Network Depth (Right).

For the Test Error vs Survival Probability at the left:
 - Both assignment rules (linear decay and uniform assignment) yield better results than constant depth.
 - Linear decay rule outperforms the uniform rule consistently.
 - Linear decay rule obtains competitive results when $p_{L}$ ranges from 0.4 to 0.8.
 - With $p_{L}=0.2$, Stochastic Depth with linear decay still performs well, while giving a 40% reduction in training time.

For the heatmap at the right:
 - Deeper networks with $p_{L}=0.5$ is better.
 - A deep enough model is necessary for stochastic depth to significantly outperform the baseline.

**Comment by NorbertZheng** [comment]
Other than dropout, we can also drop some modules by using Stochastic Depth to decrease the network depth to reduce overfitting.

**Comment by NorbertZheng** [comment]
## References
 - [2016 ECCV] [Stochastic Depth] [Deep Networks with Stochastic Depth](https://arxiv.org/abs/1603.09382).
